upickle/implicits/src-3/upickle/implicits/macros.scala

```diff
@@ -199,6 +199,54 @@ def tagKeyImpl[T](using Quotes, Type[T])(thisOuter: Expr[upickle.core.Types with
     case None => '{${thisOuter}.tagName}
   }
 
+inline def applyConstructor[T](params: Array[Any]): T = ${ applyConstructorImpl[T]('params) }
+def applyConstructorImpl[T](using quotes: Quotes, t0: Type[T])(params: Expr[Array[Any]]): Expr[T] =
+  import quotes.reflect._
+  def apply(typeApply: Option[List[TypeRepr]]) = {
+    val tpe = TypeRepr.of[T]
+    val companion: Symbol = tpe.classSymbol.get.companionModule
+    val constructorSym = tpe.typeSymbol.primaryConstructor
+    val constructorParamSymss = constructorSym.paramSymss
+
+    val (tparams0, params0) = constructorParamSymss.flatten.partition(_.isType)
+    val constructorTpe = tpe.memberType(constructorSym).widen
+
+    val rhs = params0.zipWithIndex.map {
+      case (sym0, i) =>
+        val lhs = '{$params(${ Expr(i) })}
+        val tpe0 = constructorTpe.memberType(sym0)
+
+        typeApply.map(tps => tpe0.substituteTypes(tparams0, tps)).getOrElse(tpe0) match {
+          case AnnotatedType(AppliedType(base, Seq(arg)), x)
+            if x.tpe =:= defn.RepeatedAnnot.typeRef =>
+            arg.asType match {
+              case '[t] =>
+                Typed(
+                  lhs.asTerm,
+                  TypeTree.of(using AppliedType(defn.RepeatedParamClass.typeRef, List(arg)).asType)
+                )
+            }
+          case tpe =>
+            tpe.asType match {
+              case '[t] => '{ $lhs.asInstanceOf[t] }.asTerm
+            }
+        }
+
+    }
+
+    typeApply match{
+      case None => Select.overloaded(Ref(companion), "apply", Nil, rhs).asExprOf[T]
+      case Some(args) =>
+        Select.overloaded(Ref(companion), "apply", args, rhs).asExprOf[T]
```

Comment:
just thought of this nasty example though - 
```scala
case class CustomApplyGenVararg[T](x: String, ys: T*)

  object CustomApplyGenVararg {
    def apply(x: String, ys: Int*) = new CustomApplyGenVararg[Int](x.toUpperCase, ys.map(math.abs(_)): _*)
  }
```
so there is no apply that matches the constructor

Reply:
Is there some way to say "call apply, whichever one ends up neing resolved"? That's basically what it does in Scala 2, and it works great including edge cases such as this one

Reply:
that should be what this overloaded does

Reply:
in which case, maybe you can always leave type arguments empty because the value arguments should be correct

Reply:
ok, looks like you can "try both" :)
```scala
    typeApply match {
      case None => Select.overloaded(Ref(companion), "apply", Nil, rhs, tpe).asExprOf[T]
      case Some(args) =>
        val term =
          try Select.overloaded(Ref(companion), "apply", Nil, rhs, tpe) // try in case there are no type parameters
          catch case scala.util.control.NonFatal(e) => // assume apply method needs type parameters
            Select.overloaded(Ref(companion), "apply", args, rhs, tpe)
        term.asExprOf[T]
    }
```

however it somehow still resolves to the synthetic "universal" apply method rather than the user defined one
```scala
  case class CustomApplyGen2[T](x: String, y: T)

  object CustomApplyGen2 {
    def apply(x: String, y: Int) = new CustomApplyGen2[Int](x.toUpperCase, math.abs(y))

    implicit def nodeRW[T: ReadWriter]: ReadWriter[CustomApplyGen2[T]] = macroRW[CustomApplyGen2[T]]
  }

...
      test("customApplyGen2") {
        val input = """{"x":"a","y":-102}"""
        val expected = CustomApplyGen2("A", +102)
        val result = upickle.default.read[CustomApplyGen2[Int]](input)
        assert(result == expected)
        val written = upickle.default.write(result)
        assert(written == """{"x":"A","y":102}""")
      }
...
X upickle.AdvancedTests.issues.customApplyGen2 3ms
  utest.AssertionError: result == expected
  result: upickle.AdvancedTests.CustomApplyGen2[Int] = CustomApplyGen2(a,-102)
  expected: upickle.AdvancedTests.CustomApplyGen2[Int] = CustomApplyGen2(A,102)
    utest.asserts.Asserts$.assertImpl(Asserts.scala:30)
    upickle.AdvancedTests$.$init$$$anonfun$1$$anonfun$5$$anonfun$7(AdvancedTests.scala:108)
```